### PR TITLE
Refactoring precompiles to use macros

### PR DESCRIPTION
**Pull request by asiniscalchi**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Refactored Evolution Collection precompile implementation to use macros.
- Added new precompile functions for minting, evolving, and ownership transfer.
- Implemented gas cost recording and error handling.
- Added new legacy precompile implementation.
- Added and refactored tests for Evolution Collection and legacy precompiles.
- Updated mock configuration to use new precompile structure.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>tests.rs</strong><dd><code>Refactor and add tests for Evolution Collection precompiles.</code></dd></summary>
<hr>

precompile/evolution-collection/src/tests.rs
<li>Replaced <code>MockPrecompileSet</code> with <code>LaosPrecompiles</code>.<br> <li> Updated function calls to use new <code>PrecompileCall</code> and <br><code>FactoryPrecompileCall</code>.<br> <li> Added new tests for various precompile functions.<br> <li> Refactored existing tests to use new precompile structure.<br>

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/624/files#diff-f84315de1a5d5df47a8e8fa8934f4c00bcaab0a78a3596cd581ac358071ff12a">+225/-318</a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tests.rs</strong><dd><code>Add tests for legacy evolution collection precompiles.</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

precompile/evolution-collection-legacy/src/tests.rs
<li>Added new test file for legacy evolution collection precompiles.<br> <li> Implemented tests for minting, evolving, and ownership transfer.<br>

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/624/files#diff-dfced7b514757c6fa740c3e63f1814c003466dcd06a8a6534dc82b8778f4e978">+681/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tests.rs</strong><dd><code>Add and refactor tests for Laos Evolution pallet.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pallets/laos-evolution/src/tests.rs
<li>Added tests for slot and owner to token ID conversion.<br> <li> Updated tests to use new <code>Slot</code> type.<br> <li> Refactored existing tests for minting and evolving.<br>

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/624/files#diff-6104a387e08a4830b2d78c0f033d9f5d26c16199a09d531d87a7548cf01f1efd">+43/-78</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>mock.rs</strong><dd><code>Update mock configuration for Evolution Collection precompiles.</code></dd></summary>
<hr>

precompile/evolution-collection/src/mock.rs
<li>Updated mock configuration to use new precompile structure.<br> <li> Added new type definitions for <code>PrecompileCall</code> and <br><code>FactoryPrecompileCall</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/624/files#diff-99541e36c4fc6740ed7792c9fc3516aa5ea78401db06feeeac76ba517862c187">+40/-60</a>&nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Refactor Evolution Collection precompile with macros.</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

precompile/evolution-collection/src/lib.rs
<li>Refactored precompile implementation to use macros.<br> <li> Added new precompile functions for minting, evolving, and ownership <br>transfer.<br> <li> Implemented gas cost recording and error handling.<br>

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/624/files#diff-0af32d48cea1789c31c123d4791491c2681f77280cce16aa99850142587d5a9e">+211/-252</a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>lib.rs</strong><dd><code>Implement legacy Evolution Collection precompile.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

precompile/evolution-collection-legacy/src/lib.rs
<li>Added new legacy precompile implementation.<br> <li> Implemented functions for minting, evolving, and ownership transfer.<br>

</details>
    

  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/624/files#diff-11e2c5d8342231d743d410b8a374d90b22759535bc0f045850b0cd70367f9adf">+374/-0</a>&nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: precompile/evolution-collection/src/lib.rs [hunk 1]
```diff
@@ -1,38 +1,25 @@
...
-// but WITHOUT ANY WARRANTY; without even the implied warranty of
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-// GNU General Public License for more details.
-
-// You should have received a copy of the GNU General Public License
-// along with LAOS.  If not, see <http://www.gnu.org/licenses/>.
-
-//! LAOS precompile module.
-
 #![cfg_attr(not(feature = "std"), no_std)]
```
Comment (ccubu) [inline precompile/evolution-collection/src/lib.rs:1]: This file is missing the license header

File: precompile/evolution-collection/src/tests.rs [hunk 2]
```diff
@@ -1,38 +1,18 @@
...
-// but WITHOUT ANY WARRANTY; without even the implied warranty of
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-// GNU General Public License for more details.
-
-// You should have received a copy of the GNU General Public License
-// along with LAOS.  If not, see <http://www.gnu.org/licenses/>.
-
-//! Living assets precompile tests.
-
 use super::*;
```
Comment (ccubu) [inline precompile/evolution-collection/src/tests.rs:1]: License missing

**Review by magecnion — APPROVED** [review]

**Review by ccubu — APPROVED** [review]

**Pull request merged**